### ISSUE TITLE
Kotlin DSL accessors generation could recover generics form type erasure in some cases

### ISSUE THREAD

**Issue by eskatos** [gradle/gradle#10766]
The Kotlin DSL project schema accessibility provider maps any type with generics to `Any` because of type erasure. However it should be possible to recover the generic type information in some cases like for SAM of lambda classes.

### Expected Behavior
```kotlin
// in `some` plugin
project.extensions.add("some", { name: String -> name.capitalize() })
```

```kotlin
// build.gradle.kts
plugins { `some` }
require(some("foo") == "Foo")
```

### Current Behavior
```kotlin
// in `some` plugin
project.extensions.add("some", { name: String -> name.capitalize() })
```

```kotlin
// build.gradle.kts
plugins { `some` }
// casting is required because `some` is typed `Any` because of type erasure
val casted = some as (String) -> String
require(casted("foo") == "Foo")
```

### Context
https://github.com/gradle/gradle/issues/10729
